### ISSUE TITLE
implementation of basic auth using URL userinfo

### ISSUE THREAD

**Issue by iPSAlex** [mapsforge/mapsforge#1076]
We are trying to implement a basic authentification using URL userinfo for the TileDownloader, therefore we need to use base64 encoding. The question is, is there a library that you are conform with to use in this repo?

Our approach looks like this:
```
if(userInfo != null && !userInfo.isEmpty()) {
   byte[] message = URLDecoder.decode(userInfo, "UTF-8").getBytes("UTF-8");
   String encoded = Base64.encodeToString(message, Base64.NO_WRAP);
   httpURLConnection.setRequestProperty("Authorization", "Basic " + encoded);
}
```
This is using android methods which doesn't fit to your project being plain java, are there any alternative libraries you would allow to make base64 encoding possible (i.e Apache)?

Greetings

**Comment by devemux86** [comment]
Mapsforge is platform agnostic and not heavy on dependencies that most don't need.

But those classes seem existing in both Android and Java.
Or if want more flexibility in url structure, can create a custom tile source like `OpenStreetMapMapnik` and override the `getTileUrl` method?

**Comment by iPSAlex** [comment]
To clarify, we want to edit the TileDownloader from your repo to add a basic authentification for the HttpUrlConnection. Our approach would look like this:

```
    TileBitmap downloadImage() throws IOException {
        URL url = this.downloadJob.tileSource.getTileUrl(this.downloadJob.tile);
        URLConnection urlConnection = url.openConnection();
        urlConnection.setConnectTimeout(this.downloadJob.tileSource.getTimeoutConnect());
        urlConnection.setReadTimeout(this.downloadJob.tileSource.getTimeoutRead());
        if (this.downloadJob.tileSource.getUserAgent() != null) {
            urlConnection.setRequestProperty("User-Agent", this.downloadJob.tileSource.getUserAgent());
        }
        if (this.downloadJob.tileSource.getReferer() != null) {
            urlConnection.setRequestProperty("Referer", this.downloadJob.tileSource.getReferer());
        }
        if (urlConnection instanceof HttpURLConnection) {
            HttpURLConnection httpURLConnection = (HttpURLConnection) urlConnection;
            String userInfo = url.getUserInfo();
            if(userInfo != null && !userInfo.isEmpty()) {
                byte[] message = URLDecoder.decode(userInfo, "UTF-8").getBytes("UTF-8");
                String encoded = Base64.encodeToString(message, Base64.NO_WRAP);
                httpURLConnection.setRequestProperty("Authorization", "Basic " + encoded);
            }
            httpURLConnection.setInstanceFollowRedirects(this.downloadJob.tileSource.isFollowRedirects());
        }
        InputStream inputStream = getInputStream(urlConnection);

        try {
            TileBitmap result = this.graphicFactory.createTileBitmap(inputStream, this.downloadJob.tile.tileSize,
                    this.downloadJob.hasAlpha);
            result.setExpiration(urlConnection.getExpiration());
            return result;
        } catch (CorruptedInputStreamException e) {
            // the creation of the tile bit map can fail, at least on Android,
            // when the connection is slow or busy, returning null here ensures that
            // the tile will be downloaded again
            return null;
        } finally {
            IOUtils.closeQuietly(inputStream);
        }
    }
```
The problem is that Base64 in plain Java is supported for Java 8 but not for the lower levels. An alternative would be to use DatatypeConverter from javax.xml.bind, but this is not supported for Android.

So our question is, if it is possible to add external libraries to this repo to support base64 for android as well?

Greetings

**Comment by devemux86** [comment]
You seem to need using an extra request property in url connection.

Why not pass any request property - like happens already with the others - via the tile source class?

**Comment by devemux86** [comment]
All come down to:
```java
httpURLConnection.setRequestProperty("Authorization", "Basic " + encoded);
```
which can be replaced (when implemented) with:
```java
urlConnection.setRequestProperty("Authorization", this.downloadJob.tileSource.getAuthorization());
```
That is passing a custom string as value for the request property "Authorization".
Everyone can have a custom implementation for that string, Base64 or something else.
It's not library's job to convert those strings, can be passed as parameters in final form.

> So our question is, if it is possible to add external libraries to this repo to support base64 for android as well?

There is no need (see above) and should avoid overload the library with third-party dependencies useful only for custom purposes, when there is already the flexibility work otherwise.

So next steps could be adding the `tileSource.getAuthorization()` or even better a `Map` of properties to allow any property in the future for convenience.

**Comment by iPSAlex** [comment]
Thank you for your clarification, I'll try it out!

**Comment by devemux86** [comment]
Implemented via #1081.
